File: src/main/localcli.js
'use strict';

const { spawn, execFile } = require('child_process');
const { Settings } = require('./settings');
const { parseJsonLoose } = require('./llm');

/**
 * Local developer-CLI fallback for the LLM layer.
 *
 * WHY. The four cloud providers all fail the same way in practice: a free
 * Gemini quota runs out, an HF token lacks a scope, a key is revoked. When they
 * all miss, every AI feature — translation, transliteration, mood, transcript
 * correction, per-line artist attribution — dies, even on a developer's machine
 * that has a perfectly good model a shell command away. Many devs have exactly
 * that: the Claude, Gemini, Ollama or GitHub CLIs installed and logged in. This
 * uses one of them as the last link in the chain.
 *
 * CONSENT IS REQUIRED, ALWAYS. Shelling out to someone's CLI runs their tools
 * and can spend their tokens, so this never happens silently: it stays dormant
 * until the user explicitly picks a CLI (see `setConsent`). Until then
 * `isReady()` is false and llm.js does not even offer it.
 *
 * BEST-EFFORT BY DESIGN. These are third-party CLIs whose non-interactive
 * contracts are not all stable or documented — Antigravity in particular. Every
 * adapter runs with a timeout and, on any failure, returns nothing so the
 * caller reports "all providers failed" exactly as before. A fallback that
 * hangs or throws would be worse than no fallback.
 */

/** How long to wait for a one-shot CLI call before giving up. */
const RUN_TIMEOUT_MS = 90_000;
/** How long a detection probe may take. */
const DETECT_TIMEOUT_MS = 4_000;

/**
 * The CLIs this knows how to drive, in preference order.
 *
 * Each adapter says how to check it is installed (`probe`) and how to build the
 * spawn for a one-shot prompt (`command`). The prompt always goes in on stdin,
 * never as an argument: lyric batches run to thousands of characters and a
 * command line has a hard length limit on Windows.
 */
const ADAPTERS = [
  {
    id: 'claude',
    label: 'Claude CLI',
    cmd: 'claude',
    install: 'npm install -g @anthropic-ai/claude-code',
    // `-p` is print / non-interactive: read stdin, print the reply, exit.
    // Verified end to end: returns and parses JSON.
    verified: true,
    command: () => ({ file: 'claude', args: ['-p'] }),
  },
  {
    id: 'gemini',
    label: 'Gemini CLI',
    cmd: 'gemini',
    install: 'npm install -g @google/gemini-cli',
    /*
      NOT `-p`: that flag REQUIRES the prompt as its value ("not enough
      arguments following: -p"), and our prompts are multi-line JSON that cannot
      be passed as a Windows command-line argument — cmd.exe (which shell:true
      routes through, needed for the .cmd shim) caps args near 8KB and mangles
      newlines and quotes. So it is run bare and fed the prompt on stdin, and
      relies on Gemini reading piped (non-TTY) input. Best-effort: if a build
      insists on interactive mode it times out and the chain moves on.
    */
    command: () => ({ file: 'gemini', args: [] }),
  },
  {
    id: 'ollama',
    label: 'Ollama (fully local)',
    cmd: 'ollama',
    install: 'https://ollama.com/download',
    // `ollama run <model>` reads the prompt on stdin. The model is discovered
    // at call time so this works with whatever the user has pulled.
    needsModel: 'ollama',
    command: (model) => ({ file: 'ollama', args: ['run', model || 'llama3.2'] }),
  },
  {
    id: 'copilot',
    label: 'GitHub Models (gh)',
    cmd: 'gh',
    install: 'gh extension install github/gh-models',
    // Copilot's own CLI only suggests shell commands; the gh-models extension
    // is the one that runs a chat model non-interactively, reading stdin.
    needsModel: 'gh',
    command: (model) => ({ file: 'gh', args: ['models', 'run', model || 'openai/gpt-4o-mini'] }),
  },
  {
    id: 'antigravity',
    label: 'Antigravity',
    cmd: 'antigravity',
    install: 'https://antigravity.google/',
    // Antigravity is primarily an IDE; its non-interactive contract is not
    // documented, so this is a best guess and is marked experimental in the UI.
    unverified: true,
    command: () => ({ file: 'antigravity', args: [] }),
  },
];

const settings = new Settings();

/** @returns {{consented: boolean, id: string|null}} */
function consent() {
  const raw = settings.get('localCli', null);
  if (raw && typeof raw === 'object') {
    return { consented: Boolean(raw.consented), id: raw.id || null };
  }
  return { consented: false, id: null };
}

/**
 * Record the user's choice.
 *
 * A known adapter id turns it on. The literal 'declined' is stored as a real
 * decision so the offer is not shown again — distinct from "never asked", which
 * is what an absent setting means. Anything else clears it.
 *
 * @param {string|null} id an adapter id, 'declined', or null to reset
 */
function setConsent(id) {
  if (id === 'declined') {
    settings.set('localCli', { consented: false, id: 'declined' });
    return;
  }
  const known = ADAPTERS.some((a) => a.id === id);
  settings.set('localCli', { consented: known, id: known ? id : null });
}

/** Whether a consented CLI is selected and thus usable. */
function isReady() {
  const c = consent();
  return c.consented && Boolean(c.id);
}

/** The chosen adapter, or null. */
function chosenAdapter() {
  const c = consent();
  return c.consented ? ADAPTERS.find((a) => a.id === c.id) || null : null;
}

/**
 * Is a command on PATH? Uses `where` on Windows, `which` elsewhere.
 * @param {string} cmd
 * @returns {Promise<boolean>}
 */
function onPath(cmd) {
  return new Promise((resolve) => {
    const finder = process.platform === 'win32' ? 'where' : 'which';
    execFile(finder, [cmd], { timeout: DETECT_TIMEOUT_MS, windowsHide: true }, (err, stdout) => {
      resolve(!err && Boolean((stdout || '').trim()));
    });
  });
}

/**
 * Which of the known CLIs are installed right now.
 *
 * Reports every adapter with an `installed` flag rather than only the installed
 * ones, so the UI can offer to install a missing one (see `install`).
 *
 * @returns {Promise<Array<{id: string, label: string, installed: boolean,
 *   install: string, unverified: boolean}>>}
 */
async function detect() {
  const results = await Promise.all(ADAPTERS.map(async (a) => ({
    id: a.id,
    label: a.label,
    installed: await onPath(a.cmd),
    install: a.install,
    unverified: Boolean(a.unverified),
    verified: Boolean(a.verified),
  })));
  return results;
}

/**
 * Discover a usable model name for adapters that need one.
 * @param {'ollama'|'gh'} kind
 * @returns {Promise<string|null>}
 */
function discoverModel(kind) {
  return new Promise((resolve) => {
    if (kind === 'ollama') {
      execFile('ollama', ['list'], { timeout: DETECT_TIMEOUT_MS, windowsHide: true }, (err, stdout) => {
        if (err) return resolve(null);
        // First data row, first column, is a model tag like "llama3.2:latest".
        const lines = (stdout || '').trim().split('\n').slice(1);
        const first = lines[0] && lines[0].trim().split(/\s+/)[0];
        return resolve(first || null);
      });
      return;
    }
    // gh-models: no cheap list parse worth the round trip; use a sensible default.
    resolve(null);
  });
}

/**
 * Build the single prompt string sent to a CLI.
 *
 * The CLIs have no structured-output mode, so — exactly like the Groq and HF
 * adapters in llm.js — the schema is described in the prompt and the call sites
 * validate the shape. The instruction to emit ONLY JSON is stated twice, top
 * and bottom, because a chatty model otherwise wraps it in prose.
 *
 * @param {{system: string, user: string, schema: object}} args
 * @returns {string}
 */
function buildPrompt({ system, user, schema }) {
  return `${system}

Respond with ONLY a JSON object conforming to this JSON Schema. No markdown, no
code fence, no commentary — just the JSON.

Schema:
${JSON.stringify(schema)}

Input:
${user}

Remember: output the JSON object and nothing else.`;
}

/**
 * Run the chosen CLI once and return the parsed JSON.
 *
 * Matches the `call` signature the llm.js chain expects, so it slots in as one
 * more provider. Throws on any failure — no CLI chosen, not installed, timed
 * out, non-zero exit, unparseable output — which is what makes the chain move
 * on and ultimately report that every provider failed.
 *
 * @param {{system: string, user: string, schema: object}} args
 * @returns {Promise<object>}
 */
async function call({ system, user, schema }) {
  const adapter = chosenAdapter();
  if (!adapter) throw new Error('no local CLI selected');

  let model = null;
  if (adapter.needsModel) {
    model = await discoverModel(adapter.needsModel);
    if (adapter.needsModel === 'ollama' && !model) {
      throw new Error('ollama has no models pulled (try: ollama pull llama3.2)');
    }
  }

  const { file, args } = adapter.command(model);
  const prompt = buildPrompt({ system, user, schema });

  const stdout = await new Promise((resolve, reject) => {
    let child;
    try {
      /*
        `shell: true` on Windows, because these CLIs install as `.cmd`/`.ps1`
        shims (npm global bins, gh) that `spawn` cannot execute directly —
        measured: a bare `spawn('claude', …)` fails with ENOENT even though the
        command is on PATH. The prompt goes on stdin, never in `args`, so the
        shell has nothing user-supplied to mis-split; the args are fixed
        literals plus a model tag from `ollama list`.
      */
      child = spawn(file, args, { windowsHide: true, shell: process.platform === 'win32' });
    } catch (err) {
      reject(new Error(`${adapter.id}: could not start (${err.message})`));
      return;
    }

    let out = '';
    let errText = '';
    const timer = setTimeout(() => {
      child.kill();
      reject(new Error(`${adapter.id}: timed out after ${RUN_TIMEOUT_MS / 1000}s`));
    }, RUN_TIMEOUT_MS);

    child.stdout.on('data', (d) => { out += d; });
    child.stderr.on('data', (d) => { errText += d; });
    child.on('error', (err) => {
      clearTimeout(timer);
      reject(new Error(`${adapter.id}: ${err.message}`));
    });
    child.on('close', (code) => {
      clearTimeout(timer);
      if (code !== 0) {
        reject(new Error(`${adapter.id}: exited ${code} ${errText.slice(0, 160)}`.trim()));
      } else {
        resolve(out);
      }
    });

    // Prompt on stdin: too long for a command-line argument on Windows.
    try {
      child.stdin.write(prompt);
      child.stdin.end();
    } catch (err) {
      clearTimeout(timer);
      reject(new Error(`${adapter.id}: could not write prompt (${err.message})`));
    }
  });

  return parseJsonLoose(stdout);
}

module.exports = {
  detect, setConsent, consent, isReady, call, buildPrompt, ADAPTERS,
};

File: src-tauri/src/localcli.rs
//! Local developer-CLI fallback for the LLM layer. A Rust port of `localcli.js`.
//!
//! WHY. The cloud providers all fail the same way in practice: a free Gemini
//! quota runs out, an HF token lacks a scope, a key is revoked. When they all
//! miss, every AI feature — translation, transliteration, mood, per-line
//! artist attribution — dies, even on a developer's machine that has a
//! perfectly good model a shell command away. This uses one of those (Claude,
//! Gemini, Ollama, `gh models`, Antigravity CLIs) as the last link in the
//! chain, once the user has explicitly consented (see `set_consent`).
//!
//! BEST-EFFORT BY DESIGN. These are third-party CLIs whose non-interactive
//! contracts are not all stable or documented. Every call runs with a timeout
//! and, on any failure, returns an error so the caller (`llm::convert`) falls
//! through exactly as it does for a cloud provider outage.

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::io::{Read, Write};
use std::process::{Command, Stdio};
use std::time::Duration;

const RUN_TIMEOUT: Duration = Duration::from_secs(90);
const DETECT_TIMEOUT: Duration = Duration::from_secs(4);

/// Must match `identifier` in `tauri.conf.json` — consent is stored
/// standalone (not via an `AppHandle`) so `llm::convert` can reach it from
/// deep in the provider chain without threading one through four modules.
const IDENTIFIER: &str = "com.dhruv.lyricoverlay";

struct Adapter {
    id: &'static str,
    label: &'static str,
    cmd: &'static str,
    install: &'static str,
    verified: bool,
    unverified: bool,
    needs_model: Option<&'static str>,
    /// Builds (program, args) for a one-shot non-interactive call. The prompt
    /// always goes on stdin, never in argv — lyric batches run to thousands
    /// of characters and Windows command lines have a hard length limit.
    command: fn(Option<&str>) -> (String, Vec<String>),
}

static ADAPTERS: &[Adapter] = &[
    Adapter {
        id: "claude",
        label: "Claude CLI",
        cmd: "claude",
        install: "npm install -g @anthropic-ai/claude-code",
        verified: true,
        unverified: false,
        needs_model: None,
        // `-p` is print/non-interactive: read stdin, print the reply, exit.
        command: |_| ("claude".into(), vec!["-p".into()]),
    },
    Adapter {
        id: "gemini",
        label: "Gemini CLI",
        cmd: "gemini",
        install: "npm install -g @google/gemini-cli",
        verified: false,
        unverified: false,
        needs_model: None,
        // Not `-p`: that flag requires the prompt as its own value, and ours
        // is multi-line JSON. Run bare and feed the prompt on stdin instead.
        command: |_| ("gemini".into(), vec![]),
    },
    Adapter {
        id: "ollama",
        label: "Ollama (fully local)",
        cmd: "ollama",
        install: "https://ollama.com/download",
        verified: false,
        unverified: false,
        needs_model: Some("ollama"),
        command: |model| ("ollama".into(), vec!["run".into(), model.unwrap_or("llama3.2").into()]),
    },
    Adapter {
        id: "copilot",
        label: "GitHub Models (gh)",
        cmd: "gh",
        install: "gh extension install github/gh-models",
        verified: false,
        unverified: false,
        needs_model: Some("gh"),
        command: |model| {
            ("gh".into(), vec!["models".into(), "run".into(), model.unwrap_or("openai/gpt-4o-mini").into()])
        },
    },
    Adapter {
        id: "antigravity",
        label: "Antigravity",
        cmd: "antigravity",
        install: "https://antigravity.google/",
        verified: false,
        unverified: true,
        needs_model: None,
        command: |_| ("antigravity".into(), vec![]),
    },
];

#[derive(Serialize, Deserialize, Clone, Default)]
struct Consent {
    consented: bool,
    id: Option<String>,
}

fn consent_path() -> Option<std::path::PathBuf> {
    let appdata = std::env::var_os("APPDATA")?;
    Some(std::path::PathBuf::from(appdata).join(IDENTIFIER).join("localcli.json"))
}

fn load_consent() -> Consent {
    consent_path()
        .and_then(|p| std::fs::read_to_string(p).ok())
        .and_then(|s| serde_json::from_str(&s).ok())
        .unwrap_or_default()
}

fn save_consent(c: &Consent) {
    let Some(path) = consent_path() else { return };
    if let Some(dir) = path.parent() {
        let _ = std::fs::create_dir_all(dir);
    }
    if let Ok(s) = serde_json::to_string(c) {
        let _ = std::fs::write(path, s);
    }
}

/// Record the user's choice. A known adapter id turns it on; the literal
/// `"declined"` is stored as a real decision (distinct from "never asked",
/// which is what an absent file means); anything else clears it.
pub fn set_consent(id: Option<&str>) {
    let c = match id {
        Some("declined") => Consent { consented: false, id: Some("declined".into()) },
        Some(x) if ADAPTERS.iter().any(|a| a.id == x) => Consent { consented: true, id: Some(x.to_string()) },
        _ => Consent { consented: false, id: None },
    };
    save_consent(&c);
}

pub fn consent() -> Value {
    let c = load_consent();
    json!({ "consented": c.consented, "id": c.id })
}

/// Whether a consented CLI is selected and thus usable.
pub fn is_ready() -> bool {
    let c = load_consent();
    c.consented && c.id.is_some()
}

fn chosen_adapter() -> Option<&'static Adapter> {
    let c = load_consent();
    if !c.consented {
        return None;
    }
    let id = c.id?;
    ADAPTERS.iter().find(|a| a.id == id)
}

/// Which of the known CLIs are installed right now. Reports every adapter
/// (not only installed ones) so the UI can offer an install link.
pub fn detect() -> Value {
    let results: Vec<Value> = ADAPTERS
        .iter()
        .map(|a| {
            json!({
                "id": a.id, "label": a.label, "installed": on_path(a.cmd),
                "install": a.install, "unverified": a.unverified, "verified": a.verified,
            })
        })
        .collect();
    json!(results)
}

fn on_path(cmd: &str) -> bool {
    let finder = if cfg!(windows) { "where" } else { "which" };
    matches!(run_direct(finder, &[cmd.to_string()], DETECT_TIMEOUT, None), Ok(out) if !out.trim().is_empty())
}

/// Discover a usable model name for adapters that need one.
fn discover_model(kind: &str) -> Option<String> {
    if kind != "ollama" {
        // gh-models: no cheap list parse worth the round trip; a default is used.
        return None;
    }
    let out = run_direct("ollama", &["list".to_string()], DETECT_TIMEOUT, None).ok()?;
    // First data row (after the header), first column: a tag like "llama3.2:latest".
    let first_line = out.trim().lines().nth(1)?;
    first_line.split_whitespace().next().map(String::from)
}

/// Build the single prompt string sent to a CLI. These have no structured-
/// output mode, so — like the Groq/HF adapters in `llm.rs` — the schema is
/// described in the prompt and the caller validates the shape.
fn build_prompt(system: &str, user: &str, schema: &Value) -> String {
    format!(
        "{system}\n\n\
Respond with ONLY a JSON object conforming to this JSON Schema. No markdown, no\n\
code fence, no commentary — just the JSON.\n\n\
Schema:\n{schema}\n\n\
Input:\n{user}\n\n\
Remember: output the JSON object and nothing else."
    )
}

/// Run the chosen CLI once and return the parsed JSON. Matches the shape
/// `llm::convert`'s chain expects, so it slots in as one more provider.
pub fn call(system: &str, user: &str, schema: &Value) -> Result<Value, String> {
    let adapter = chosen_adapter().ok_or("no local CLI selected")?;

    let mut model = None;
    if let Some(kind) = adapter.needs_model {
        model = discover_model(kind);
        if kind == "ollama" && model.is_none() {
            return Err("ollama has no models pulled (try: ollama pull llama3.2)".into());
        }
    }

    let (file, args) = (adapter.command)(model.as_deref());
    let prompt = build_prompt(system, user, schema);
    let out = run_cli(&file, &args, RUN_TIMEOUT, &prompt).map_err(|e| format!("{}: {e}", adapter.id))?;
    crate::llm::parse_json_loose(&out)
}

/// Windows npm-global CLIs install as `.cmd`/`.ps1` shims that a direct
/// CreateProcess (what `Command::new` does) cannot execute even when the
/// command is on PATH — measured, matches the same problem `localcli.js`
/// worked around with `shell: true`. Route through `cmd /C` there; spawn
/// directly everywhere else.
fn run_cli(file: &str, args: &[String], timeout: Duration, stdin_data: &str) -> Result<String, String> {
    if cfg!(windows) {
        let mut full_args = vec!["/C".to_string(), file.to_string()];
        full_args.extend(args.iter().cloned());
        run_direct("cmd", &full_args, timeout, Some(stdin_data))
    } else {
        run_direct(file, args, timeout, Some(stdin_data))
    }
}

/// Spawn a process, optionally write `stdin_data` then close stdin, and wait
/// up to `timeout` — killing and erroring out past it. stdout/stderr are
/// drained on their own threads throughout (not just after exit) so a CLI
/// that writes more than fits in one pipe buffer before exiting can't
/// deadlock the wait.
fn run_direct(program: &str, args: &[String], timeout: Duration, stdin_data: Option<&str>) -> Result<String, String> {
    let mut cmd = Command::new(program);
    cmd.args(args).stdin(Stdio::piped()).stdout(Stdio::piped()).stderr(Stdio::piped());
    #[cfg(windows)]
    {
        use std::os::windows::process::CommandExt;
        const CREATE_NO_WINDOW: u32 = 0x0800_0000;
        cmd.creation_flags(CREATE_NO_WINDOW);
    }

    let mut child = cmd.spawn().map_err(|e| format!("could not start: {e}"))?;

    let mut stdin = child.stdin.take();
    let data = stdin_data.map(String::from);
    std::thread::spawn(move || {
        if let (Some(si), Some(d)) = (stdin.as_mut(), data) {
            let _ = si.write_all(d.as_bytes());
        }
        // `si` (and `stdin`) drop here either way, closing the pipe.
    });

    let mut stdout = child.stdout.take();
    let out_handle = std::thread::spawn(move || {
        let mut buf = String::new();
        if let Some(so) = stdout.as_mut() {
            let _ = so.read_to_string(&mut buf);
        }
        buf
    });
    let mut stderr = child.stderr.take();
    let err_handle = std::thread::spawn(move || {
        let mut buf = String::new();
        if let Some(se) = stderr.as_mut() {
            let _ = se.read_to_string(&mut buf);
        }
        buf
    });

    let start = std::time::Instant::now();
    let status = loop {
        if let Ok(Some(status)) = child.try_wait() {
            break Some(status);
        }
        if start.elapsed() > timeout {
            let _ = child.kill();
            let _ = child.wait();
            break None;
        }
        std::thread::sleep(Duration::from_millis(50));
    };

    let out = out_handle.join().unwrap_or_default();
    let err = err_handle.join().unwrap_or_default();

    match status {
        None => Err(format!("timed out after {}s", timeout.as_secs())),
        Some(s) if s.success() => Ok(out),
        Some(s) => Err(format!("exited {:?}: {}", s.code(), err.chars().take(160).collect::<String>())),
    }
}
